### PR TITLE
Optimize the implementation not to use any external function.

### PR DESCRIPTION
**Pull request by sjrd**
Here is the implementation of the `fround` polyfill used by Scala.js 1.9.0+ when targeting ECMAScript 5.1, adapted for variable names to match the previous variable names in this codebase. The inspiration for using Veltkamp's splitting for the normal form case form came from the polyfill in core-js, which seems to also be the inspiration for the previous algorithm in this codebase. I have however reconstructed a proof with references that it is indeed correct. I designed the algorithm (if it can be called that) for the subnormal form case.

I intentionally break two eslint rules for the sake of performance:

* using `+x` instead of `$Number(x)`
* hard-coding the constants right inside the algorithm.

If that is not welcome in this repository, or misguided, I am happy to revert to using `$Number` and/or declaring the magic constants in `var`s as before.

---

The special code path for zeros and NaNs is removed. They fall through as the general case.

We avoid `$Number` by using the asm.js idiom `v = +x`. We avoid `abs` by using `s * v`. This causes `-0` to fall through as is, but it turns out that it has the desired effect in the end.

We optimize the subnormal path to use only one multiplication and one division.

We dispatch the overflow case from the normal and subnormal cases ahead of time, which allows to more easily justify the correctness of the algorithm.

We hard-code the magical constants to avoid having to look up external identifiers, ensuring that immediate values are used in the generated code.

Finally, and perhaps most importantly, we add significant documentation about the algorithms for the normal and subnormal paths, with---if not proof---at least an argument and references of why they are correct.

### CODE REVIEW

**Comment by sjrd** [comment]
I'm confused by the CI errors on (IIUC) Node.js 1-4. They don't seem related to my changes, but the `main` branch builds successfully. Did I do something wrong when removing the dependency on `is-nan`?

File: implementation.js [hunk 1]
```diff
@@ -1,42 +1,107 @@
...
-	var result = a - (a - mod);
-	if (result > BINARY_32_MAX_VALUE || isNaN(result)) {
+	/*
+	 * Overflow case.
+	 *
+	 * The magic number 3.4028235677973366E38 is the threshold from which fround
+	 * causes an overflow, resulting in Infinity.
+	 * This code path is also used when the input is already an Infinity.
+	 */
+	if (mod >= 3.4028235677973366E38) { // overflow-threshold
```
Comment (nicolo-ribaudo) [inline implementation.js:23]: The old implementation used `3.4028234663852886e+38`, which is lower than this number. I have no idea about what is expected, but could you add tests for numbers around these values?

We already have one:
https://github.com/es-shims/Math.fround/blob/be55c4c233d19494399f866bfe90d8891cc4fbf7/test/tests.js#L48-L49

It would probably be good to test both the new magic number and `3.4028235677973362e+38`, which is the float that comes right before it.

**Review by nicolo-ribaudo — APPROVED** [review]
LGTM, leaving to @ljharb the questions about code style and CI.

**Comment by nicolo-ribaudo** [comment]
Do you have a microbenchmark that shouws the perf diff?

**Comment by ljharb** [comment]
How does hardcoding the constants in the algorithm improve performance?

Any performance-motivated change should definitely include benchmarks (in the PR is fine). In particular, I'd like comparisons of:
 - the current approach
 - the PR's approach
 - the PR's approach, but caching constants at module level and using `Number()` over `+`
 
Additionally, performance is only a sensible concept to discuss in the context of a specific implementation. This package supports basically every version of Firefox, Safari, IE, Edge, Chrome, node, and many others - so we'd want to compare performance in all of these.

File: implementation.js (hunk 1 above)
Comment (sjrd) [reply implementation.js]: > The old implementation used `3.4028234663852886e+38`, which is lower than this number.

The old implementation used the max-binary32 value. The new implementation refers to the *midpoint* between max-binary32 and Infinity in binary32.


> It would probably be good to test both the new magic number and `3.4028235677973362e+38`, which is the float that comes right before it.

The new magic number was already tested in "returns infinity for large numbers". I have added a test for the float just below it. I have also added tests around the subnormal-normal boundary.

**Comment by sjrd** [comment]
I don't have *micro*-benchmarks, but I do have [a benchmark of a raytracer](https://github.com/sjrd/scalajs-benchmarks/tree/302017cd9cdfaadd7d5bd3cc93bb76b957786039/tracerFloat/shared/src/main/scala/org/scalajs/benchmark/tracer) where every single floating point value is a `Float`, and therefore every floating point operation goes through `fround`. Here are the results:

| Which variant | Mean running time (lower is better) | Standard error the Mean |
|---|---|---|
| Current approach | 4828.66 | 5.86 |
| PR approach with caching | 3802.77 | 3.74 |
| PR as is | 3525.61 | 3.67 |

> How does hardcoding the constants in the algorithm improve performance?

I guess it's easier for the engine's JIT to compile them to immediate values, rather than loading them from memory or having to protect the immediates with deoptimization guards.

> Additionally, performance is only a sensible concept to discuss in the context of a specific implementation. This package supports basically every version of Firefox, Safari, IE, Edge, Chrome, node, and many others - so we'd want to compare performance in all of these.

True. Here the performance measurements are done on a Linux Ubuntu machine with Node.js 14.14.0. This is probably not representative of all the possible versions. But I would say that an implementation that uses 0 external function call has a higher chance of being compiled in a consistent way across as many engines as possible.
